### PR TITLE
Add w3c context propagation tests

### PR DESCRIPTION
**Pull request by malafeev**
It's already done for Python and JS.
Although most of the tests pass there are some which fail.
Because execution of the tests is not part of the circleci workflow it doesn't affect build.
```
----------------------------------------------------------------------
Ran 40 tests in 1.220s

FAILED (failures=3, skipped=6)
```

### CODE REVIEW

File: integration_tests/tracecontext/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# OpenTelemetry W3C Context Propagation tests
```
Comment (jkwatson) [inline integration_tests/tracecontext/README.md:1]: Can you add some notes in here about the environmental requirements to have this be runnable? It looks like you'll need to be on some sort of system that supports a bourne shell, and have python3/pip3 installed.

**Comment by jkwatson** [comment]
What tests are failing? Can we add Issues to address them?

File: integration_tests/tracecontext/README.md (hunk 1 above)
Comment (anuraaga) [reply integration_tests/tracecontext/README.md:1]: It'd be great if we can push a docker image with this test environment. I wouldn't block this PR on it, but then we'd even be able to run it as part of the build.

Comment (jkwatson) [reply integration_tests/tracecontext/README.md:1]: I would have to guess that there's a stock ubuntu image out there with python3/pip3 already ready-to-go.

**Comment by malafeev** [comment]
Failed tests:
```
======================================================================
ERROR: test_tracestate_empty_header (__main__.TraceContextTest)
harness sends a request with empty tracestate header
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 516, in test_tracestate_empty_header
    self.assertEqual(tracestate['foo'], '1')
  File "/Users/malafes/projects-opentelemetry/malafeev-opentelemetry-java/integration_tests/tracecontext/target/trace-context/test/tracecontext/tracestate.py", line 32, in __getitem__
    return self._traits[key]
KeyError: 'foo'

======================================================================
ERROR: test_tracestate_multiple_headers_different_keys (__main__.TraceContextTest)
harness sends a request with multiple tracestate headers, each contains different set of keys
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 531, in test_tracestate_multiple_headers_different_keys
    self.assertTrue(str(tracestate).index('bar=2') < str(tracestate).index('rojo=1'))
ValueError: substring not found

======================================================================
FAIL: test_traceparent_duplicated (__main__.TraceContextTest)
harness sends a request with two traceparent headers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 143, in test_traceparent_duplicated
    self.assertNotEqual(traceparent.trace_id.hex(), '12345678901234567890123456789011')
AssertionError: '12345678901234567890123456789011' == '12345678901234567890123456789011'

======================================================================
FAIL: test_traceparent_version_0x00 (__main__.TraceContextTest)
harness sends an invalid traceparent with extra trailing characters
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 194, in test_traceparent_version_0x00
    self.assertNotEqual(traceparent.trace_id.hex(), '12345678901234567890123456789012')
AssertionError: '12345678901234567890123456789012' == '12345678901234567890123456789012'

======================================================================
FAIL: test_traceparent_version_0xff (__main__.TraceContextTest)
harness sends an invalid traceparent with version 255 (0xff)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 224, in test_traceparent_version_0xff
    self.assertNotEqual(traceparent.trace_id.hex(), '12345678901234567890123456789012')
AssertionError: '12345678901234567890123456789012' == '12345678901234567890123456789012'

======================================================================
FAIL: test_traceparent_version_illegal_characters (__main__.TraceContextTest)
harness sends an invalid traceparent with illegal characters in version
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 234, in test_traceparent_version_illegal_characters
    self.assertNotEqual(traceparent.trace_id.hex(), '12345678901234567890123456789012')
AssertionError: '12345678901234567890123456789012' == '12345678901234567890123456789012'

----------------------------------------------------------------------
Ran 40 tests in 1.021s

FAILED (failures=4, errors=2, skipped=6)
```

File: integration_tests/tracecontext/README.md (hunk 1 above)
Comment (malafeev) [reply integration_tests/tracecontext/README.md:1]: updated readme and circleci config to run in docker image

File: .circleci/config.yml [hunk 2]
```diff
@@ -107,13 +107,36 @@ jobs:
...
+            - ~/.gradle
+          key: gradle-w3c-{{ checksum "build.gradle" }}
 
 workflows:
   version: 2
   build_and_test:
     jobs:
       - build
       - java11
+      - integration-tests
```
Comment (jkwatson) [inline .circleci/config.yml:139]: If we have failing tests, won't this fail every build until we get them fixed?

Comment (malafeev) [reply .circleci/config.yml:140]: get docker image with python3 and pip3 made my day crazy... still not be able to do that

Comment (jkwatson) [reply .circleci/config.yml:140]: we can hold on that for a follow-up if you want. I'd be happy getting the tests in as a first step, then automating the running of them as a second step.

File: .circleci/config.yml [hunk 3]
```diff
@@ -107,13 +107,36 @@ jobs:
...
             - ~/.gradle
           key: gradle-{{ checksum "build.gradle" }}
+# Integration tests
+  integration-tests:
+    environment:
+      # Configure the JVM and Gradle to avoid OOM errors
+      _JAVA_OPTIONS: "-Xmx1g"
+      GRADLE_OPTS: "-Dorg.gradle.daemon=false -Dorg.gradle.workers.max=2"
+    docker:
+      - image: openkbs/jdk-mvn-py3
```
Comment (malafeev) [inline .circleci/config.yml:117]: this docker image has pip3, but output shows that pip3 not found.
how it's possible?

Comment (jkwatson) [reply .circleci/config.yml:117]: maybe it's not on the default PATH?

Comment (malafeev) [reply .circleci/config.yml]: no, something different

File: .circleci/config.yml (hunk 2 above)
Comment (malafeev) [reply .circleci/config.yml]: I removed automation

File: integration_tests/tracecontext/README.md (hunk 1 above)
Comment (malafeev) [reply integration_tests/tracecontext/README.md:1]: circleci config reverted

**Comment by jkwatson** [comment]
Do you have any thoughts on how we can keep it up-to-date if there are changes on the w3c side?

**Comment by malafeev** [comment]
> Do you have any thoughts on how we can keep it up-to-date if there are changes on the w3c side?

that's a good question, right now commit is hardcoded (same in python and js).

**Comment by malafeev** [comment]
@jkwatson I created 3 issues for failed tests: 
- https://github.com/open-telemetry/opentelemetry-java/issues/1674
- https://github.com/open-telemetry/opentelemetry-java/issues/1675
- https://github.com/open-telemetry/opentelemetry-java/issues/1676

**Review by jkwatson — APPROVED** [review]

**Pull request merged**